### ISSUE TITLE
Test strict mode, nightly browsers

### ISSUE THREAD

**Issue by zcorpan** [scottjehl/picturefill#489]
#478 could have been discovered earlier if we were testing picturefill with WebKit trunk (which supports `currentSrc` but not `HTMLPictureElement`; shipped Safari does not support `currentSrc`) and `"use strict"`. Evidently this was not done, but sites are using picturefill in strict mode.

* Run the tests both with and without `"use strict"`.
* Run the tests in nightly browsers as well to catch problems early.

(Possibly also quirks mode/standards mode and IE's dozens of compat modes, if you will.)

**Comment by Wilto** [comment]
Yeah, for sure—we’re well past the point of keeping track of support in our heads, which we could do for a little while there.

Any thoughts on automating this?

@baloneysandwiches, you were working on an Intern/Sauce Labs setup for PF3 testing, yeah? Any luck so far?

**Comment by aFarkas** [comment]
Question: does safari nightly support sizes but not picture?

**Comment by baloneysandwiches** [comment]
I am pretty sold on intern/saucelabs for automation. I test-drove it on the parser:

https://github.com/baloneysandwiches/parse-srcset/blob/master/tests/unit/ps.js

I haven't dealt with a grunt-based setup, but it can be done.

This is on my radar. Sorry, my plate just keeps filling back up the moment I clear it. It's initially just a matter of restructuring existing tests. Then adding.

**Comment by patrickkettner** [comment]
I just finished getting saucelabs browser testing in modernizr.

saucelabs is waaaaay slower on updating their browser support compared to browserstack. That being said, the general UI/X of testing on saucelabs is quite a bit better. My notes comparing the two from a few months back

1. Browserstack UI doesn't show the difference between finished, failed, and passed. Its all a green check mark. Sauce has passed/failed/timeedout labeled on each test
2. BrowserStack test and command line UI doesn't show anything other than the number of tests passed and failed, sauce shows the actual failures as well as the stacks. 
3. BrowserStack don't currently support accessing another server, only static files. That means if you do something like instrument Istanbul coverage reporting, you can't use it. Sauce supports any endpoint in general.

If you want to always test the latest and greatest, then you will probably want to try to make browserstack work.

**Comment by patrickkettner** [comment]
oh, also, if you want to support testing on sauce via PRs, you will want to look into something like [savage](https://github.com/twbs/savage)

**Comment by csnover** [comment]
I was notified about this thread and just wanted to butt in with a few fun facts! :)

>  Browserstack UI doesn't show the difference between finished, failed, and passed. Its all a green check mark. Sauce has passed/failed/timeedout labeled on each test

This is not correct, you need to use a test system that provides this information. Intern tells BrowserStack success/failure and it is reflected in the BrowserStack UI:

![screen shot 2015-04-23 at 13 04 15](https://cloud.githubusercontent.com/assets/229244/7303918/6643230e-e9b9-11e4-82e8-5b0fac176b9b.png)
![screen shot 2015-04-23 at 13 04 23](https://cloud.githubusercontent.com/assets/229244/7303921/67a2756a-e9b9-11e4-83fc-986c49f8e05b.png)

> BrowserStack test and command line UI doesn't show anything other than the number of tests passed and failed, sauce shows the actual failures as well as the stacks.

This might be true of whatever test system BrowserStack wrote themselves, but it’s all just a matter of using the right test system. Intern provides all this information regardless of which cloud provider you use, so Intern + BrowserStack gives you all the same information as Intern + Sauce Labs or Intern + TestingBot or Intern + Selenium.

> BrowserStack don't currently support accessing another server, only static files. That means if you do something like instrument Istanbul coverage reporting, you can't use it. Sauce supports any endpoint in general.

BrowserStack integration with Intern is first class and allows you to interact with any local or remote server, including code instrumentation (which Intern gives you for free out-of-the-box).

Best,

**Comment by patrickkettner** [comment]
@csnover thank you so much for the clarification! All of my notes were using their browserstack-runner tool, at that time.

**Comment by patrickkettner** [comment]
@csnover do you have an image showing build failures? My memory is foggy, but I think my issue with the UI was the large green completed checkmark still showed up even if the build failed. Could be wrong, however

**Comment by csnover** [comment]
@patrickkettner The second screenshot in my last post shows a single test build failure. The exclamation point shows up the same in the sidebar, as well in the build matrixes for any failures (I don’t have any matrix test failures to screenshot though).

**Comment by patrickkettner** [comment]
the mixed results was what I was referring to, it would be great to see

**Comment by zcorpan** [comment]
@aFarkas
> Question: does safari nightly support sizes but not picture?

Yes.

**Comment by zcorpan** [comment]
Does browserstack support nightly (or other pre-release channels)?

**Comment by patrickkettner** [comment]
they have dev, neither support nightly

On Fri, Apr 24, 2015 at 2:55 AM, Simon Pieters <notifications@github.com>
wrote:

> Does browserstack support nightly (or other pre-release channels)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scottjehl/picturefill/issues/489#issuecomment-95824168>
> .
>



-- 
patrick

**Comment by mike-engel** [comment]
Quick ping to @jefflembeck and @baloneysandwiches to see if they've made any progress, or even have time still to take this on.

**Comment by aFarkas** [comment]
Just some note, I tested the 3.0 branch with Safari nightly (-> green).

**Comment by Wilto** [comment]
Bump. Any luck so far?
